src/EditorFeatures/Test/MetadataAsSource/MetadataAsSourceTests.cs

```diff
@@ -2366,5 +2369,45 @@ void M([|D|]&lt;int&gt; lambda)
             var metadataAsSourceFile = await context.GenerateSourceAsync(navigationSymbol);
             context.VerifyResult(metadataAsSourceFile, expected);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
+        public async Task TestFailedDecompilation()
+        {
+            var exportProvider = ExportProviderCache.GetOrCreateExportProviderFactory(TestExportProvider.EntireAssemblyCatalogWithCSharpAndVisualBasic.WithPart(typeof(ThrowingDecompiledSourceService))).CreateExportProvider();
+            var throwingService = exportProvider.GetExportedValue<ThrowingDecompiledSourceService>();
+            var expectedException = throwingService.Exception;
+            var lazyFailureText = new LazyToString(() => "// " + expectedException.ToString().Replace("\n", "\n// "));
+
+            var metadataSource = "public class C {}";
+            var symbolName = "C";
+
+            // Using the decompiler forces the language to C#, so the expected output is the same in both cases
+            IFormattable expected = $@"#region {FeaturesResources.Assembly} ReferencedAssembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
+// {CodeAnalysisResources.InMemoryAssembly}
+#endregion
+
+public class [|C|]
+{{
+    public C();
+}}
+{lazyFailureText}";
+
+            await GenerateAndVerifySourceAsync(metadataSource, symbolName, LanguageNames.CSharp, expected, allowDecompilation: true, exportProvider: exportProvider);
+            await GenerateAndVerifySourceAsync(metadataSource, symbolName, LanguageNames.VisualBasic, expected, allowDecompilation: true, exportProvider: exportProvider);
+        }
+
+        [Export]
+        [ExportLanguageService(typeof(IDecompiledSourceService), LanguageNames.CSharp, ServiceLayer.Host)]
+        [Shared]
+        [PartNotDiscoverable]
+        private sealed class ThrowingDecompiledSourceService : IDecompiledSourceService
+        {
+            public Exception Exception { get; } = new NotImplementedException();
```

Comment:
It seems accessing this property should throw if AddSourceToAsync() hasn't been called yet; I get the laziness in the tests is working around this, but anybody changing the test harness and causes a more eager evaluation is going to get a baseline failure rather than a more explicit "this shouldn't have been called yet".

Reply:
Changed the test to `Lazy<string>` and evaluated this inside the lambda.